src/ray/core_worker/reference_count_test.cc

```diff
@@ -1556,6 +1561,89 @@ TEST(DistributedReferenceCountTest, TestDuplicateBorrower) {
+// inner_id = ray.get(outer_id)[0]
+// return_id2 = owner_task2.remote(inner_id)
+//
+TEST(DistributedReferenceCountTest, TestForeignOwner) {
+  auto caller = std::make_shared<MockWorkerClient>("1");
+  auto owner = std::make_shared<MockWorkerClient>("2");
+  auto foreign_owner = std::make_shared<MockWorkerClient>("3");
+
+  //
+  // Phase 1 -- submit and execute owner_task1()
+  //
+  // Caller submits a task.
+  auto return_id = caller->SubmitTaskWithArg(ObjectID::Nil());
+  // Task returns inner_id as its return value.
+  auto inner_id = ObjectID::FromRandom();
+  owner->PutWithForeignOwner(inner_id, foreign_owner->address_);
+  rpc::WorkerAddress addr(caller->address_);
+  auto refs = owner->FinishExecutingTask(ObjectID::Nil(), return_id, &inner_id, &addr);
+  ASSERT_TRUE(refs.empty());
+  ASSERT_TRUE(owner->rc_.HasReference(inner_id));
+  // Caller receives the owner's message, but inner_id is still in scope
+  // because caller has a reference to return_id.
+  caller->HandleSubmittedTaskFinished(return_id, ObjectID::Nil(),
+                                      {{return_id, {inner_id}}});
+
+  //
+  // Phase 2 -- submit and execute owner_task2(x)
+  //
+  auto return_id2 = caller->SubmitTaskWithArg(return_id);
+  caller->rc_.RemoveLocalReference(return_id, nullptr);
+  ASSERT_TRUE(owner->rc_.HasReference(inner_id));
+  caller->rc_.RemoveLocalReference(return_id2, nullptr);
+  // Owner receives a reference to inner_id. It still has a reference when
+  // the task returns.
+  owner->ExecuteTaskWithArg(return_id, inner_id, owner->address_);
+  auto refs2 = owner->FinishExecutingTask(return_id, return_id);
+  // owner merges ref count into the caller.
+  caller->HandleSubmittedTaskFinished(return_id2, return_id, {}, owner->address_, refs2);
+  ASSERT_FALSE(owner->rc_.HasReference(return_id));
+  ASSERT_FALSE(caller->rc_.HasReference(return_id));
+  ASSERT_FALSE(owner->rc_.HasReference(return_id2));
+  ASSERT_FALSE(caller->rc_.HasReference(return_id2));
+  ASSERT_TRUE(owner->rc_.HasReference(inner_id));
+
+  //
+  // Phase 3 -- foreign owner gets ref removed information.
+  //
+  // Emulate ref removed callback.
+  ReferenceCounter::ReferenceTableProto refs_proto;
+  owner->rc_.GetAndClearLocalBorrowersInternal(inner_id, /*for_ref_removed=*/true,
+                                               &refs_proto);
+  ASSERT_TRUE(owner->rc_.HasReference(inner_id));
+
+  // This is the key assert checking the info is still available.
+  ASSERT_EQ(refs_proto[0].stored_in_objects().size(), 1);
+
+  owner->rc_.RemoveLocalReference(inner_id, nullptr);
+  owner->rc_.RemoveLocalReference(inner_id, nullptr);
+  owner->rc_.RemoveLocalReference(inner_id, nullptr);
+  owner->rc_.RemoveLocalReference(return_id, nullptr);
+  owner->rc_.RemoveLocalReference(return_id2, nullptr);
+  ASSERT_FALSE(owner->rc_.HasReference(inner_id));
+  ASSERT_FALSE(foreign_owner->rc_.HasReference(inner_id));
+  ASSERT_FALSE(caller->rc_.HasReference(inner_id));
```

Comment:
@stephanie-wang need a bit a bit of help figuring out why this last assert is failing.

Reply:
I'm not sure about this line, but the line above should fail because the foreign_owner has to send another WaitForRefRemoved RPC to caller (can trigger the handler with `caller->FlushBorrowerCallbacks()`).

Reply:
It turns out this was a real reference leak (had to change the logic to not return the borrower reference in the foreign owner case).